File: C++/3. Longest Substring Without Repeating Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int visit[128];
        fill(visit, visit + 128, -1); //visit[] store the index of s[i]
        int start = 0; // start = the beginning index of temp substring
        int max_str = 0; // max_str = the length of temp substring
        int res = 0;// res = the length of longest substring

        for (int i = 0; i < s.size(); i++) {
            //if duplicate
            if (visit[s[i]] != -1) {
                if (visit[s[i]] >= start ) //only care about the duplicate character after start
                {
                    start = visit[s[i]] + 1;
                    max_str =  i - start ;
                }
            }
            max_str += 1;
            visit[s[i]] = i;
            if (max_str > res)
                res = max_str;
        }

        return res;
    }
};

File: Python/3. Longest Substring Without Repeating Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        start = 0
        sub = ""
        max_len = 0
        for i in range(len(s)):
            start = sub.find(s[i], 0, len(sub))

            if start == -1:  # substring don't contain s[i]
                sub += s[i]
            else:  # s[i] in substring
                if(start+1 < len(sub)):
                    # dropout string before s[i] in substring
                    sub = sub[(start+1):]
                else:
                    sub = ""
                sub += s[i]
            if len(sub) > max_len:
                max_len = len(sub)

        return max_len
